ledger/src/snapshot_utils.rs

```diff
@@ -138,6 +143,95 @@ where
     }
 }
 
+pub fn serialize_snapshot_data_file<F>(
+    data_file_path: &Path,
+    maximum_file_size: u64,
+    mut serializer: F,
+) -> Result<u64>
+where
+    F: FnMut(&mut BufWriter<File>) -> Result<()>,
+{
+    let data_file = File::create(data_file_path)?;
+    let mut data_file_stream = BufWriter::new(data_file);
+    serializer(&mut data_file_stream)?;
+    data_file_stream.flush()?;
+
+    let consumed_size = data_file_stream.seek(SeekFrom::Current(0))?;
+    if consumed_size > maximum_file_size {
+        let error_message = format!(
+            "too large snapshot data file to serialize: {:?} has {} bytes",
+            data_file_path, consumed_size
+        );
+        return Err(get_io_error(&error_message));
+    }
+    Ok(consumed_size)
+}
+
+pub fn deserialize_snapshot_data_file<F, T>(
+    data_file_path: &Path,
+    maximum_file_size: u64,
+    mut deserializer: F,
+) -> Result<T>
+where
+    F: FnMut(&mut BufReader<File>) -> Result<T>,
+{
+    let file_size = fs::metadata(&data_file_path)?.len();
+
+    if file_size > maximum_file_size {
+        let error_message = format!(
+            "too large snapshot data file to deserialize: {:?} has {} bytes",
+            data_file_path, file_size
+        );
+        return Err(get_io_error(&error_message));
+    }
+
+    let data_file = File::open(data_file_path)?;
+    let mut data_file_stream = BufReader::new(data_file);
+
+    let ret = deserializer(&mut data_file_stream)?;
+
+    let consumed_size = data_file_stream.seek(SeekFrom::Current(0))?;
+
+    if file_size != consumed_size {
+        let error_message = format!(
+            "invalid snapshot data file: {:?} has {} bytes, however consumed {} bytes to deserialize",
+            data_file_path, file_size, consumed_size
+        );
+        return Err(get_io_error(&error_message));
+    }
+
+    Ok(ret)
+}
+
+macro_rules! serialize_snapshot_data_file_with_metrics {
```

Comment:
My first moderate sized macro! :)

Reply:
Due to `inc_new_counter_info!` requirement that its first argument must be `static`-friendly, I was forced to write a macro rather than a wrapper `fn`.

Reply:
🤔I see how it reduces the need to duplicate most of the `datapoint_info!()` and `inc_new_counter_info!()` macros between two call sites.  However this minor duplication seems preferable to the added complexity of a large macro such as this.

Reply:
Ok. Now I'm getting used to the balance of complexity/duplication of codebase for Soalna. :) De-macro-ed d06a0ba6de60dcd611ff7713f7269e69da66c6df